File: apps/cortensor-claim-check/lib/cortensorClient.ts
import { CORTENSOR_ROUTER, LLM_PARAMETERS } from '@/config/app';

interface CortensorChoice {
  message?: { content?: string };
  text?: string;
}

interface CortensorCompletionResponse {
  id?: string;
  task_id?: number | string;
  choices?: CortensorChoice[];
  output?: string;
  text?: string;
  error?: { message?: string };
}

export interface CortensorTaskResponse {
  assigned_miners?: string[];
  created_at?: number;
  finished_at?: number;
  results?: {
    data?: string[];
    miners?: string[];
  };
  status?: number;
}

const completionsBase = CORTENSOR_ROUTER.completionsUrl.replace(/\/$/, '');
const completionsEndpointForSession = (sessionId: number) =>
  (/\/\d+$/.test(completionsBase) ? completionsBase : `${completionsBase}/${sessionId}`);

const headers = () => {
  if (!CORTENSOR_ROUTER.apiKey) {
    throw new Error('Missing CORTENSOR_API_KEY environment variable.');
  }

  return {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${CORTENSOR_ROUTER.apiKey}`,
  } satisfies HeadersInit;
};

const normalizeContent = (payload: CortensorCompletionResponse) => {
  const choice = payload.choices?.[0];
  return choice?.message?.content ?? choice?.text ?? payload.output ?? payload.text ?? '';
};

export async function requestCortensorCompletion(
  prompt: string,
  options?: { sessionId?: number; promptType?: number; clientReference?: string },
): Promise<{
  content: string;
  raw: CortensorCompletionResponse;
  taskId: number;
}> {
  const targetSessionId = options?.sessionId ?? LLM_PARAMETERS.sessionId;
  const body = {
    prompt,
    prompt_type: options?.promptType ?? LLM_PARAMETERS.promptType,
    prompt_template: '',
    stream: false,
    timeout: LLM_PARAMETERS.timeoutSeconds,
    precommit_timeout: 90,
    client_reference: options?.clientReference ?? 'claimcheck-oracle',
    max_tokens: LLM_PARAMETERS.maxTokens,
    temperature: 0.7,
    top_p: 0.95,
    top_k: 40,
    presence_penalty: 0,
    frequency_penalty: 0,
  };

  const response = await fetch(completionsEndpointForSession(targetSessionId), {
    method: 'POST',
    headers: headers(),
    body: JSON.stringify(body),
    cache: 'no-store',
  });

  const json = (await response.json()) as CortensorCompletionResponse;

  if (!response.ok) {
    const message = json.error?.message ?? `Cortensor responded with status ${response.status}.`;
    throw new Error(message);
  }

  const content = normalizeContent(json).trim();
  const taskIdValue = json.task_id;
  const taskId = typeof taskIdValue === 'number' ? taskIdValue : Number(taskIdValue);

  if (!Number.isFinite(taskId)) {
    throw new Error('Cortensor completion did not return a task_id.');
  }

  if (!content) {
    throw new Error('Cortensor completion returned empty content.');
  }

  return { content, raw: json, taskId };
}

const taskEndpointBaseForSession = (sessionId: number) => `${CORTENSOR_ROUTER.baseUrl}/api/v1/tasks/${sessionId}`;

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

async function fetchTaskOnce(taskId: number | string, sessionId: number): Promise<CortensorTaskResponse> {
  const response = await fetch(`${taskEndpointBaseForSession(sessionId)}/${taskId}`, {
    method: 'GET',
    headers: headers(),
    cache: 'no-store',
  });

  if (!response.ok) {
    const message = `Failed to fetch task ${taskId}: ${response.statusText}`;
    throw new Error(message);
  }

  return (await response.json()) as CortensorTaskResponse;
}

export async function fetchTaskSnapshot(
  taskId: number | string,
  options?: { sessionId?: number },
): Promise<CortensorTaskResponse> {
  const sessionId = options?.sessionId ?? LLM_PARAMETERS.sessionId;
  return fetchTaskOnce(taskId, sessionId);
}

export async function waitForTaskResults(
  taskId: number | string,
  options?: { maxAttempts?: number; delayMs?: number; sessionId?: number },
) {
  const { maxAttempts = 8, delayMs = 1500, sessionId = LLM_PARAMETERS.sessionId } = options ?? {};
  let attempt = 0;
  let lastError: Error | undefined;

  while (attempt < maxAttempts) {
    attempt += 1;
    try {
      const task = await fetchTaskOnce(taskId, sessionId);
      const outputs = task.results?.data ?? [];
      const minerCount = (task.results?.miners ?? task.assigned_miners ?? []).length;
      const isComplete =
        outputs.length > 0 &&
        (minerCount === 0
          ? task.status === 4 || Boolean(task.finished_at)
          : outputs.length >= minerCount);

      if (isComplete) {
        return task;
      }
      lastError = undefined;
    } catch (error) {
      lastError = error instanceof Error ? error : new Error('Unknown error polling Cortensor task.');
    }

    await sleep(delayMs);
  }

  if (lastError) {
    throw lastError;
  }

  throw new Error(`Timed out waiting for Cortensor task ${taskId} to finish.`);
}

File: apps/cortensor-recaptgbot/src/cortensor_client.py
from __future__ import annotations

import asyncio
from typing import Any, Dict, Optional

import aiohttp


class CortensorClient:
    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: int = 90,
        session_id: Optional[int] = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.session_id = session_id
        self._client: Optional[aiohttp.ClientSession] = None

    async def __aenter__(self) -> "CortensorClient":
        self._client = aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=self.timeout)
        )
        return self

    async def __aexit__(self, *exc: Any) -> None:
        if self._client:
            await self._client.close()

    @property
    def client(self) -> aiohttp.ClientSession:
        if not self._client:
            raise RuntimeError("Client session not initialized")
        return self._client

    async def summarize(self, prompt: str, **kwargs) -> str:
        # Accepts extra Cortensor params via kwargs
        payload: Dict[str, Any] = {
            "prompt": prompt,
            "prompt_type": kwargs.get("prompt_type", 0),
            "prompt_template": kwargs.get("prompt_template", ""),
            "stream": False,
            "timeout": self.timeout,
            "precommit_timeout": kwargs.get("precommit_timeout", 90),
            "client_reference": kwargs.get("client_reference", "user-request-123"),
            "max_tokens": kwargs.get("max_tokens", 1024),
            "temperature": kwargs.get("temperature", 0.7),
            "top_p": kwargs.get("top_p", 0.95),
            "top_k": kwargs.get("top_k", 40),
            "presence_penalty": kwargs.get("presence_penalty", 0.0),
            "frequency_penalty": kwargs.get("frequency_penalty", 0.0),
        }
        # Use sessionId in URL if present, else in body
        if self.session_id is not None:
            url = f"{self.base_url}/api/v1/completions/{self.session_id}"
        else:
            url = f"{self.base_url}/api/v1/completions"
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        async with self.client.post(url, headers=headers, json=payload) as resp:
            data = await resp.json(content_type=None)
            if resp.status >= 300:
                raise RuntimeError(
                    f"Cortensor error {resp.status}: {data!r}"
                )
            return self._extract_text(data)

    @staticmethod
    def _extract_text(data: Dict[str, Any]) -> str:
        # Try OpenAI-style response first
        choices = data.get("choices")
        if isinstance(choices, list) and choices:
            text = choices[0].get("text") or choices[0].get("message", {}).get("content")
            if text:
                return text

        # Fallbacks used by some Router builds
        if "completion" in data and isinstance(data["completion"], str):
            return data["completion"]
        if "data" in data and isinstance(data["data"], str):
            return data["data"]

        return str(data)

    async def healthcheck(self) -> Dict[str, Any]:
        headers = {"Authorization": f"Bearer {self.api_key}"}
        url = f"{self.base_url}/api/v1/info"
        async with self.client.get(url, headers=headers) as resp:
            return await resp.json(content_type=None)


async def summarize_text(*args: Any, **kwargs: Any) -> str:
    async with CortensorClient(*args, **kwargs) as client:
        return await client.summarize(kwargs.get("prompt", ""))


async def check_router(*args: Any, **kwargs: Any) -> Dict[str, Any]:
    async with CortensorClient(*args, **kwargs) as client:
        return await client.healthcheck()
